src/Contractors.jl

```diff
@@ -48,192 +36,104 @@ derivative(f) = x-> f(Taylor1([x,1.],1))[1]
-			@debug "Krawczyk candidate", Nx
 		end
-		 
 		x = intersect(x, Nx)
-
-		if (x_old == x && diam(x) < ϵ) || isempty(x) 
-			return x
+		if !(x_old == x)  # Newton / Krawczyk was effective and improved the interval
+			continue
 		end
-
-		#= if x_old == x
-			@debug fif x_old == x
-			@debug fm
-			a, b = Interval(x.lo), Interval(x.hi)
-			fa, fb = f(a), f(b)
-
-			if 0 ∈ fm
-				# if 0 belongs to the image of the midpoint, 
-				# we implement  
-				# a bisection search is better
-
-				@debug "0 is in fm"
-				rint = Interval(radius(x))
-				
-				# rand has average 1/2, so in average this is a bisection step
-
-				x_cand_l = x_mid-1/14*rint
-				@debug x_cand_l, f(x_cand_l)
-				x_cand_r = x_mid+3/4*rint
-				@debug x_cand_r, f(x_cand_r)
-				
-				left = Interval(x.lo) 
-				right = Interval(x.hi)
-
-				if !(0 ∈ f(x_cand_l)) 
-					left = x_cand_l
-				end
-				if !(0 ∈ f(x_cand_r))
-					right = x_cand_r
-				end
-
-				
-				x = hull(left, right)#	enl = 1
-				
-				@debug "After bisection step", x
-				if x == x_old
-					return x
-				end
-			else
-				if !(0 ∈ hull(fa, fm))
-					x = Interval(x_mid, x.hi)
-				else 
-					x = Interval(x.lo, x_mid)
-				end
-			end
+		# otherwise, let us switch to a zero-th order method based on bisection
+		# fm has already been computed, let us use it
+
+		# Notice that this bisection strategy works also when f is not monotonic outside [a,b]; 
+		# see https://github.com/JuliaDynamics/RigorousInvariantMeasures.jl/issues/140#issuecomment-1413541452 for a proof
+		if y ∉ hull(y1, fm)
+			x = Interval(x_mid.hi, x.hi)
+			continue
+		end
+		if y ∉ hull(fm, y2)
+			x = Interval(x.lo, x_mid.lo)
+			continue
 		end
-			a, b = Interval(x.lo), Interval(x.hi)
-			fa, fb = f(a), f(b)
-
-			if 0 ∈ fm
-				# if 0 belongs to the image of the midpoint, 
-				# we implement  
-				# a bisection search is better
-
-				@debug "0 is in fm"
-				rint = Interval(radius(x))
-				
-				# rand has average 1/2, so in average this is a bisection step
-
-				x_cand_l = x_mid-1/14*rint
-				@debug x_cand_l, f(x_cand_l)
-				x_cand_r = x_mid+3/4*rint
-				@debug x_cand_r, f(x_cand_r)
-				
-				left = Interval(x.lo) 
-				right = Interval(x.hi)
-
-				if !(0 ∈ f(x_cand_l)) 
-					left = x_cand_l
-				end
-				if !(0 ∈ f(x_cand_r))
-					right = x_cand_r
-				end
 
-				
-				x = hull(left, right)#	enl = 1
-				
-				@debug "After bisection step", x
-				if x == x_old
-					return x
-				end
-			else
-				if !(0 ∈ hull(fa, fm))
-					x = Interval(x_mid, x.hi)
-				else 
-					x = Interval(x.lo, x_mid)
-				end
+		# this may still fail if y ≈ fm or if the function is computed in a very inaccurate way (relative to the size of `x`)
+		# in this case, we try chopping off smaller and smaller parts of `x` at both ends
+		# If we cannot even cut off 1/2^6th of the interval it's time to give up.
+		for k = 2:6
+			c = x.lo + diam(x) / 2^k
+			if y ∉ hull(y1, f(Interval(c)))
+				x = Interval(c, x.hi)
+				continue
```

Comment:
hmm, this `continue` is a mistake. :) My idea was that if we succeed to chop off anything from the interval then we can continue with the next iteration of the *outer* for loop, the `for i in 1:max_iter`. Since we have shrunk the interval, we can go back to trying Newton on the new interval and hope that it works. I'll fix it, thanks for noticing it.